docs/migration/wsc54/php.md

```diff
@@ -31,3 +31,114 @@ Additionally, to now also support deleting phrases with this package installatio
 Note that when deleting phrases, the category does not have to be specified because phrase identifiers are unique globally.
 
 !!! warning "Mixing the old structure and the new structure is not supported and will result in an error message during the import!"
+
+## Events
+
+Historically events were tightly coupled with a single class, with the event object being an object this class, expecting the event listener to consume public properties and method of the event object.
+The `$parameters` array was introduced due to limitations of this pattern, avoiding moving all the values that might be of interest to the event listener into the state of the object.
+Events were still tightly coupled with the class that fired the event and using the opaque parameters array prevented IDEs from assisting with autocompletion and typing.
+
+WoltLab Suite 5.5 introduces the concept of dedicated, reusable event classes.
+Any newly introduced event will receive a dedicated class, implementing the `wcf\system\event\IEvent` interface.
+These event classes may be fired from multiple locations, making them reusable to convey that a conceptual action happened, instead of a specific class doing something.
+An example could be a user logging in.
+Instead of listening on a the login form being submitted and the Facebook login action successfully running, an event listener `UserLoggedIn` might be fired whenever a user logs in, no matter how the login is performed.
+
+Additionally these dedicated event classes will benefit from full IDE support.
+All the relevant values may be stored as real properties on the event object.
+
+Previously:
+
+```php
+$parameters = [
+    'value' => \random_int(1, 1024),
+];
+
+EventHandler::getInstance()->fireAction($this, 'foo', $parameters);
+```
+
+{jinja{ codebox(
+    language="php",
+    title="lib/system/event/listener/FooDumpListener.class.php",
+    contents="""
+<?php
+
+namespace wcf\system\event\listener;
+
+use wcf\form\FooForm;
+
+final class FooDumpListener implements IParameterizedEventListener
+{
+    /**
+     * @inheritDoc
+     * @param FooForm $eventObj
+     */
+    public function execute($eventObj, $className, $eventName, array &$parameters)
+    {
+        var_dump($parameters['value']);
+    }
+}
+"""
+) }}
+
+Now:
+
+```
+EventHandler::getInstance()->fire(new \wcf\system\foo\Event(\random_int(1, 1024)));
```

Comment:
There should be an explicit statement above on where (i.e. in which namespace) the new event classes should be placed.

Reply:
My intention was placing these event classes where they logically fit based off the namespace instead of having an event dumping ground.

e.g.:
- `\wcf\system\user\authentication\UserLoggedIn` 
- `\wcf\system\email\EmailSend`
- `\wbb\data\post\PostDeleted`

All of them possibly using either an Event suffix or an additional `\event` in the namespace, similar to exceptions.

Reply:
I do not like the `Event` suffix being sprayed everywhere, it adds quite a lot of noise, where it really is not required. I'm somewhat leaning towards the additional `\event` namespace, so these classes are kept separated.

```php
use \wcf\system\user\authentication\event\UserLoggedIn;

new UserLoggedIn();
```
vs.

```php
use \wcf\system\user\authentication\UserLoggedInEvent;

new UserLoggedInEvent();
```

Reply:
I've now improved the example with better choices for event / class names.